// mdx-m3-viewer/src/utils/mappeddata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MappedData = void 0;
const file_1 = require("../parsers/slk/file");
const file_2 = require("../parsers/ini/file");
/**
 * A structure that holds mapped data from INI and SLK files.
 *
 * In the case of SLK files, the first row is expected to hold the names of the columns.
 */
class MappedData {
    constructor(buffer) {
        this.map = {};
        if (buffer) {
            this.load(buffer);
        }
    }
    /**
     * Load data from an SLK file or an INI file.
     *
     * Note that this may override previous properties!
     */
    load(buffer) {
        if (buffer.startsWith('ID;')) {
            const file = new file_1.default();
            file.load(buffer);
            const rows = file.rows;
            const header = rows[0];
            const map = this.map;
            for (let i = 1, l = rows.length; i < l; i++) {
                const row = rows[i];
                // DialogueDemonBase.slk has an empty row.
                if (row) {
                    let name = row[0];
                    // DialogueDemonBase.slk also has rows containing only a single underline.
                    if (name && name !== '_') {
                        name = name.toLowerCase();
                        if (!map[name]) {
                            map[name] = {};
                        }
                        const mapped = map[name];
                        for (let j = 0, k = header.length; j < k; j++) {
                            let key = header[j];
                            // UnitBalance.slk doesn't define the name of one column.
                            if (key === undefined) {
                                key = `column${j}`;
                            }
                            mapped[`${key}`] = row[j];
                        }
                    }
                }
            }
        }
        else {
            const file = new file_2.IniFile();
            file.load(buffer);
            const sections = file.sections;
            const map = this.map;
            for (const [row, properties] of sections.entries()) {
                if (!map[row]) {
                    map[row] = {};
                }
                const mapped = map[row];
                for (const [name, property] of properties) {
                    mapped[name] = property;
                }
            }
        }
    }
    getRow(key) {
        return this.map[key.toLowerCase()];
    }
    getProperty(key, name) {
        return this.map[key.toLowerCase()][name];
    }
    setRow(key, values) {
        this.map[key.toLowerCase()] = values;
    }
    findRow(key, expectedValue) {
        for (const row of Object.values(this.map)) {
            if (row[key] === expectedValue) {
                return row;
            }
        }
        return;
    }
}
exports.MappedData = MappedData;
