// typescript-to-lua/dist/transformation/visitors/jsx/jsx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsxVisitors = exports.formatJSXStringValueLiteral = void 0;
const ts = require("typescript");
const lua = require("../../../LuaAST");
const literal_1 = require("../literal");
const xhtml_1 = require("./xhtml");
const annotations_1 = require("../../utils/annotations");
function findAnnotationByType(node, fileAnnotation) {
    const annotation = annotations_1.getFileAnnotations(node.getSourceFile()).get(fileAnnotation);
    return annotation === null || annotation === void 0 ? void 0 : annotation.args[0];
}
function getExpressionFromOption(option) {
    const [first, second] = option.split(".");
    return second === undefined
        ? lua.createIdentifier(first)
        : lua.createTableIndexExpression(lua.createIdentifier(first), lua.createStringLiteral(second));
}
function getJsxFactory(node, context) {
    var _a, _b;
    const option = (_b = (_a = findAnnotationByType(node, annotations_1.AnnotationKind.Jsx)) !== null && _a !== void 0 ? _a : context.options.jsxFactory) !== null && _b !== void 0 ? _b : "React.createElement";
    return getExpressionFromOption(option);
}
function getJsxFragmentName(node, context) {
    var _a, _b;
    const option = (_b = (_a = findAnnotationByType(node, annotations_1.AnnotationKind.JsxFrag)) !== null && _a !== void 0 ? _a : context.options.jsxFragmentFactory) !== null && _b !== void 0 ? _b : "React.Fragment";
    return getExpressionFromOption(option);
}
/*
The following 3 functions for jsx text processing modified from sucrase (https://github.com/alangpierce/sucrase), which
is published with the MIT licence:

The MIT License (MIT)

Copyright (c) 2012-2018 various contributors (see AUTHORS)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
const HEX_NUMBER = /^[\da-fA-F]+$/;
const DECIMAL_NUMBER = /^\d+$/;
/**
 * Turn the given jsxText string into a JS string literal. Leading and trailing
 * whitespace on lines is removed, except immediately after the open-tag and
 * before the close-tag. Empty lines are completely removed, and spaces are
 * added between lines after that.
 *
 * We trim the start and end of each line and remove blank lines.
 */
function formatJSXTextLiteral(text) {
    let result = "";
    let whitespace = "";
    let isInInitialLineWhitespace = false;
    let seenNonWhitespace = false;
    for (let i = 0; i < text.length; i++) {
        const c = text[i];
        if (c === " " || c === "\t" || c === "\r") {
            if (!isInInitialLineWhitespace) {
                whitespace += c;
            }
        }
        else if (c === "\n") {
            whitespace = "";
            isInInitialLineWhitespace = true;
        }
        else {
            if (seenNonWhitespace && isInInitialLineWhitespace) {
                result += " ";
            }
            result += whitespace;
            whitespace = "";
            if (c === "&") {
                const { entity, newI } = processEntity(text, i + 1);
                i = newI - 1;
                result += entity;
            }
            else {
                result += c;
            }
            seenNonWhitespace = true;
            isInInitialLineWhitespace = false;
        }
    }
    if (!isInInitialLineWhitespace) {
        result += whitespace;
    }
    return result;
}
/**
 * Format a string in the value position of a JSX prop.
 *
 * Use the same implementation as convertAttribute from
 * babel-helper-builder-react-jsx.
 */
// changes from sucrase: no multi-line flattening of prop strings in typescript.
function formatJSXStringValueLiteral(text) {
    let result = "";
    for (let i = 0; i < text.length; i++) {
        const c = text[i];
        if (c === "&") {
            const { entity, newI } = processEntity(text, i + 1);
            result += entity;
            i = newI - 1;
        }
        else {
            result += c;
        }
    }
    return result;
}
exports.formatJSXStringValueLiteral = formatJSXStringValueLiteral;
/**
 * Modified from jsxReadString in Babylon.
 */
function processEntity(text, indexAfterAmpersand) {
    let str = "";
    let count = 0;
    let entity;
    let i = indexAfterAmpersand;
    while (i < text.length && count++ < 10) {
        const ch = text[i];
        i++;
        if (ch === ";") {
            if (str.startsWith("#")) {
                if (str[1] === "x") {
                    str = str.substr(2);
                    if (HEX_NUMBER.test(str)) {
                        entity = String.fromCodePoint(Number.parseInt(str, 16));
                    }
                }
                else {
                    str = str.substr(1);
                    if (DECIMAL_NUMBER.test(str)) {
                        entity = String.fromCodePoint(Number.parseInt(str, 10));
                    }
                }
            }
            else {
                entity = xhtml_1.XHTMLEntities[str];
            }
            break;
        }
        str += ch;
    }
    if (!entity) {
        return { entity: "&", newI: indexAfterAmpersand };
    }
    return { entity, newI: i };
}
// end functions copied from sucrase
function processJsxText(jsxText) {
    const text = formatJSXTextLiteral(jsxText.text);
    if (text === "")
        return undefined;
    return ts.factory.createStringLiteral(text);
}
const charCodes = {
    a: 0x61,
    z: 0x7a,
};
// how typescript does it
function isIntrinsicJsxName(escapedName) {
    const name = escapedName;
    const ch = name.charCodeAt(0);
    return (ch >= charCodes.a && ch <= charCodes.z) || name.includes("-") || name.includes(":");
}
function transformTagName(name, context) {
    if (ts.isIdentifier(name) && isIntrinsicJsxName(name.escapedText)) {
        return lua.createStringLiteral(ts.idText(name), name);
    }
    else {
        return context.transformExpression(name);
    }
}
function transformJsxChildren(children, context) {
    if (!children)
        return undefined;
    return children
        .map(child => {
        if (ts.isJsxText(child)) {
            return processJsxText(child);
        }
        if (ts.isJsxExpression(child)) {
            return child.expression;
        }
        return child;
    })
        .filter(child => child !== undefined)
        .map(child => context.transformExpression(child));
}
function createJsxFactoryCall(tagName, props, tsChildren, tsOriginal, context) {
    const transformedChildren = transformJsxChildren(tsChildren, context);
    const jsxFactory = getJsxFactory(tsOriginal, context);
    const args = [tagName];
    if (props) {
        args.push(props);
    }
    if (transformedChildren && transformedChildren.length > 0) {
        if (!props) {
            args.push(lua.createNilLiteral());
        }
        args.push(...transformedChildren);
    }
    return lua.createCallExpression(jsxFactory, args, tsOriginal);
}
function transformJsxOpeningLikeElement(node, children, context) {
    const tagName = transformTagName(node.tagName, context);
    const props = node.attributes.properties.length !== 0 ? literal_1.transformJsxAttributes(node.attributes, context) : undefined;
    return createJsxFactoryCall(tagName, props, children, node, context);
}
const transformJsxElement = (node, context) => transformJsxOpeningLikeElement(node.openingElement, node.children, context);
const transformSelfClosingJsxElement = (node, context) => transformJsxOpeningLikeElement(node, undefined, context);
const transformJsxFragment = (node, context) => {
    const tagName = getJsxFragmentName(node, context);
    return createJsxFactoryCall(tagName, undefined, node.children, node, context);
};
exports.jsxVisitors = {
    [ts.SyntaxKind.JsxElement]: transformJsxElement,
    [ts.SyntaxKind.JsxSelfClosingElement]: transformSelfClosingJsxElement,
    [ts.SyntaxKind.JsxFragment]: transformJsxFragment,
};
